sklearn/preprocessing/data.py

```diff
@@ -1767,135 +1755,185 @@ class OneHotEncoder(BaseEstimator, TransformerMixin):
-            except (ValueError, TypeError):
-                raise TypeError("Wrong type for parameter `n_values`. Expected"
-                                " 'auto', int or array of ints, got %r"
-                                % type(X))
-            if n_values.ndim < 1 or n_values.shape[0] != X.shape[1]:
-                raise ValueError("Shape mismatch: if n_values is an array,"
-                                 " it has to be of shape (n_features,).")
-
-        self.n_values_ = n_values
-        n_values = np.hstack([[0], n_values])
-        indices = np.cumsum(n_values)
-        self.feature_indices_ = indices
+        self._n_features = n_features
 
-        column_indices = (X + indices[:-1]).ravel()
-        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
-                                n_features)
-        data = np.ones(n_samples * n_features)
-        out = sparse.coo_matrix((data, (row_indices, column_indices)),
-                                shape=(n_samples, indices[-1]),
-                                dtype=self.dtype).tocsr()
+        self.label_encoders_ = [LabelEncoder() for i in range(n_features)]
 
-        if self.n_values == 'auto':
-            mask = np.array(out.sum(axis=0)).ravel() != 0
-            active_features = np.where(mask)[0]
-            out = out[:, active_features]
-            self.active_features_ = active_features
+        if self.n_values != 'auto':
+            warnings.warn('The parameter `n_values` is deprecated, use the'
+                          'parameter `classes_` instead and specify the '
+                          'expected values for each feature')
 
-        return out if self.sparse else out.toarray()
+            if isinstance(self.n_values, numbers.Integral):
+                if (np.max(X, axis=0) >= self.n_values).any():
+                    raise ValueError("Feature out of bounds for n_values=%d"
+                                     % self.n_values)
+                n_values = np.empty(n_features, dtype=np.int)
+                n_values.fill(self.n_values)
+            else:
+                try:
+                    n_values = np.asarray(self.n_values, dtype=int)
+                except (ValueError, TypeError):
+                    raise TypeError("Wrong type for parameter `n_values`."
+                                    " Expected 'auto', int or array of ints,"
+                                    "got %r" % type(X))
+                if n_values.ndim < 1 or n_values.shape[0] != X.shape[1]:
+                    raise ValueError("Shape mismatch: if n_values is an array,"
+                                     " it has to be of shape (n_features,).")
+
+            self.classes = [np.arange(n, dtype=np.int) for n in n_values]
+
+        for i in range(n_features):
+            le = self.label_encoders_[i]
+            if self.classes == 'auto':
+                le.fit(X[:, i])
+            else:
+                le.classes_ = np.array(self.classes[i])
```

Comment:
Btw, `X` is useless in this case right?

Should we make `X` optional and/or check that all categories in `X[:, i]` are present in `classes[i]`?